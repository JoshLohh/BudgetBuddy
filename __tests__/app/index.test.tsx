/**
 * Tests for HomeScreen (index.tsx)
 * - Validates rendering of welcome message and navigation links
 * - Checks rendering of HelloWave and layout structure
 * - Mocks navigation and theming components for isolated testing
 */

import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import HomeScreen from '../../app/index';

// Mock child dependencies:
jest.mock('@/components/HelloWave', () => ({
  HelloWave: () => <></>, // render empty fragment (mock animation)
}));

jest.mock('@/components/ThemedText', () => {
  const React = require('react');
  const { Text } = require('react-native');
  return { ThemedText: ({ children, ...props }: any) => <Text {...props}>{children}</Text> };
});

jest.mock('@/components/ThemedView', () => {
  const React = require('react');
  const { View } = require('react-native');
  return { ThemedView: ({ children, ...props }: any) => <View {...props}>{children}</View> };
});

jest.mock('@/components/Spacer', () => {
  const React = require('react');
  const { View } = require('react-native');
  return ({ testID }: any) => <View testID={testID || 'spacer'} />;
});

jest.mock('expo-router', () => {
    const React = require('react');
    const { Text } = require('react-native');
  
    const router = {
      push: jest.fn(),
    };
  
    // Simulate Link: onPress will call router.push(href)
    const Link = ({ children, href, ...props }: any) => (
      <Text
        accessibilityRole="link"
        onPress={() => {
          if (href) router.push(href);
        }}
        {...props}
      >
        {children}
      </Text>
    );
  
    return {
      Link,
      router,
    };
});
  

describe('HomeScreen', () => {
  it('renders welcome message', () => {
    const { getByText } = render(<HomeScreen />);
    expect(getByText('Welcome to BudgetBuddy')).toBeTruthy();
  });

  it('renders navigation links/buttons for Login, Register, Profile', () => {
    const { getByText } = render(<HomeScreen />);
    expect(getByText('Login')).toBeTruthy();
    expect(getByText('Register')).toBeTruthy();
    expect(getByText('Profile')).toBeTruthy();
  });

  it('renders the HelloWave component', () => {
    const { UNSAFE_getByType } = render(<HomeScreen />);
    const HelloWave = require('@/components/HelloWave').HelloWave;
    expect(UNSAFE_getByType(HelloWave)).toBeTruthy();
  });

  it('includes Spacer components for layout spacing', () => {
    const { getAllByTestId } = render(<HomeScreen />);
    const spacers = getAllByTestId('spacer');
    expect(spacers.length).toBeGreaterThan(0);
  });

  it('calls router.push when profile link is pressed (simulate interaction)', () => {
    const { getByText } = render(<HomeScreen />);
    const { router } = require('expo-router');
    const profileLink = getByText('Profile');
    fireEvent.press(profileLink);
    expect(router.push).toHaveBeenCalledWith('/profile/profile');
  });

  // Add snapshot test for structural changes
  it('matches snapshot', () => {
    const { toJSON } = render(<HomeScreen />);
    expect(toJSON()).toMatchSnapshot();
  });
});
