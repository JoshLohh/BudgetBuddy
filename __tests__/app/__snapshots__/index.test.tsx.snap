// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeScreen matches snapshot 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <ViewManagerAdapter_ExpoImage
    containerViewRef={"[React.ref]"}
    contentFit="contain"
    contentPosition={
      {
        "left": "50%",
        "top": "50%",
      }
    }
    height={200}
    nativeViewRef={"[React.ref]"}
    onError={[Function]}
    onLoad={[Function]}
    onLoadStart={[Function]}
    onProgress={[Function]}
    placeholder={[]}
    source={
      [
        {
          "testUri": "../../../assets/images/logo.png",
        },
      ]
    }
    style={
      {
        "height": 200,
        "width": 200,
      }
    }
    transition={null}
    width={200}
  />
  <View
    testID="spacer"
  />
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "gap": 8,
        "justifyContent": "center",
      }
    }
  >
    <Text
      style={
        {
          "textAlign": "center",
        }
      }
      type="title"
    >
      Welcome to BudgetBuddy
    </Text>
  </View>
  <View
    testID="spacer"
  />
  <View
    testID="spacer"
  />
  <View
    style={
      {
        "alignItems": "center",
        "gap": 8,
        "marginBottom": 8,
      }
    }
  >
    <Text
      accessibilityRole="link"
      onPress={[Function]}
    >
      <Text
        type="link"
      >
        Login
      </Text>
    </Text>
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "gap": 8,
        "marginBottom": 8,
      }
    }
  >
    <Text
      accessibilityRole="link"
      onPress={[Function]}
    >
      <Text
        type="link"
      >
        Register
      </Text>
    </Text>
  </View>
  <View
    style={
      {
        "alignItems": "center",
        "gap": 8,
        "marginBottom": 8,
      }
    }
  >
    <Text
      accessibilityRole="link"
      onPress={[Function]}
    >
      <Text
        type="link"
      >
        Profile
      </Text>
    </Text>
  </View>
</View>
`;
